### ISSUE TITLE
Specify freenet module and ceno node interface

### ISSUE THREAD

**Issue by redwire** [equalitie/ceno#5]
@billdoran has asked me to start outlining some code that describes how the freenet module will interface with ceno node, and I have done some (most?) of that work now in [my new branch](https://github.com/redwire/ceno/commits/freenetinterface), but I have simplified the interaction in a way that seems more natural to me.

Bill suggested that we were going to have things work as follows:

1. The ceno-node receives a request from the client browser and makes an HTTP request to the freenet http-handler
2. The freenet http handler checks the cache to see if the page has a cached bundle it can serve
3. If one exists, the cached bundle is served
4. If not, the freenet module responds to the ceno node module with a unique ID
    1. The ceno node then requests the bundle again
    2. The Ceno node this time serves the bundle, which we expect has been created and using the unique ID pass it to the freenet module for storage

-----

All of step 4 here seems redundant to me, and so I have simplified the process to the following:

1. The freenet module receives a request from the client browser and makes an HTTP request to the Ceno node
2. The Ceno node checks the cache to see if the page has a cached bundle it can serve
3. If one exists, the cached bundle is served
4. If not, the Ceno node immediately calls upon the bundler, supplying it with the request and response objects associated with the request, and has it serve a newly created and cached bundle itself

If there is a particular reason we required the extra requests in the prior sequence, please make a case for them here.  I'd like to make sure I have the interaction understood properly. It would be great to receive more information that I can use to flesh out the `freenet.js` file, too.

**Comment by vmon** [comment]
First, let  have two principal in mind:
- both Ceno-node and the freenet plugin will break into two part, the ceno-client side and the ceno-bridge side. For now we are only doing simulation to have simplest version ready.
- We want to have ceno node to be persistent, while be able to write plugins for different storage and anonymous communication channel. For now, freenet is both the storage and channel, but we wouldn't mind to use MaidSAFE, i2p, Tor, GnuNet/Secushare etc as storage/communication channel). So we would like to plug the plugin out without changing to the ceno-node and put something else in as listener to port 3090.

Having said that, let get back to your suggestion. What is not working with your step 2 of your protocol:

> 2. The Ceno node checks the cache to see if the page has a cached bundle it can serve

Ceno node has no way of checking the cache, because the cache is stored on freenet (having in mind that freenet is a big stretched hard drive).

Just to point out that in reality, there is step 3.5, that ceno-client use anonymous channel to request the page from ceno-bridge. (that can happen through freenet, tor, i2p, etc, but we are sticking to freenet for now for simplicity). Then ceno-bridge bundles and ask freenet to store the page into cache (freenet).

In summary, I suggest that you write something to replace the functionality of

F's = require('fs.extra')

With http request to a plugin listening localhost:3090, so for example fs.existsSync(this.bodyCachePath) in:

var bodyExists = fs.existsSync(this.bodyCachePath);

will make a http request and see if the plugin say that bundle exists in freenet or not.

**Comment by redwire** [comment]
I am opening an issue with a proposal for how to do all of this, so I am closing this issue since the task of outlining a spec is technically done (if not perfect).
